conda_forge_tick/migrators/version.py

```diff
@@ -1,136 +1,281 @@
-        flags=re.M,
-    )
-    r_url_pat = re.compile(
-        r"^(\s*)(-)?(\s*)url:\s*(?:(#.*)?\[([^\[\]]+)\])?(?(4)[^\(\)]*?)\n(\1(?(2) \3)  -.*\n?)*",  # noqa
-        flags=re.M,
-    )
-    r_urls = re.compile(r"\s*-\s*(.+?)(?:#.*)?$", flags=re.M)
 
-    migrator_version = 0
+def _is_r_url(url: str):
+    if (
+        "cran.r-project.org/src/contrib" in url or
+        "cran_mirror" in url
+    ):
+        return True
+    else:
+        return False
 
-    # TODO: replace these types with a namedtuple so that it is clear what this is
-    def find_urls(self, text: str) -> List[Tuple[Union[str, List[str]], str, str]]:
-        """Get the URLs and platforms in a meta.yaml."""
-        urls: List[Tuple[Union[str, List[str]], str, str]] = []
-        for m in self.url_pat.finditer(text):
-            urls.append((m.group(4), m.group(6), m.group()))
-        for m in self.r_url_pat.finditer(text):
-            if m is not None:
-                r = self.r_urls.findall(m.group())
-                urls.append((r, m.group(2), m.group()))
-        return urls
-
-    def get_hash_patterns(
-        self,
-        filename: str,
-        urls: List[Tuple[Union[str, List[str]], str, str]],
-        hash_type: str,
-    ) -> Sequence[Tuple[str, str, str]]:
-        """Get the patterns to replace hash for each platform."""
-        pats: MutableSequence[Tuple[str, str, str]] = []
-        checksum_names = [
-            "hash_value",
-            "hash",
-            "hash_val",
-            "sha256sum",
-            "checksum",
-            hash_type,
-        ]
-        for url, platform, line in urls:
-            if isinstance(url, list):
-                for u in url:
-                    u = u.strip("'\"")
-                    try:
-                        hash_ = hash_url(u, hash_type)
-                        break
-                    except urllib.error.HTTPError:
-                        continue
-                else:
-                    raise ValueError("Could not determine hash from recipe!")
+
+def _has_r_url(curr_val: Any):
+    if isinstance(curr_val, collections.abc.MutableSequence):
+        for i in range(len(curr_val)):
+            return _has_r_url(curr_val[i])
+    elif isinstance(curr_val, collections.abc.MutableMapping):
+        for key in _gen_key_selector(curr_val, 'url'):
+            return _has_r_url(curr_val[key])
+    elif isinstance(curr_val, str):
+        return _is_r_url(curr_val)
+    else:
+        return False
+
+
+def _compile_all_selectors(cmeta: Any, src: str):
+    selectors = [None]
+    for key in cmeta.jinja2_vars:
+        if CONDA_SELECTOR in key:
+            selectors.append(key.split(CONDA_SELECTOR)[1])
+    for key in src:
+        if CONDA_SELECTOR in key:
+            selectors.append(key.split(CONDA_SELECTOR)[1])
+    return set(selectors)
+
+
+def _try_url_and_hash_it(url: str, hash_type: str):
+    print("    downloading url:", url)
```

Comment:
Probably want to change these out for logger.debug

Reply:
done. ruamel.yaml cleans lint so I added a linted version of the input yaml

Reply:
done.